projects/observability/src/shared/components/topology/d3/layouts/graph-layout.ts

```diff
@@ -0,0 +1,76 @@
+import { RenderableTopology, RenderableTopologyNode, TopologyCoordinates, TopologyEdge, TopologyNode } from "../../topology";
+
+export class GraphLayout {
+  public layout(topology: RenderableTopology<TopologyNode, TopologyEdge>): void {
+
+    const levelToNodesMap = new Map<number, RenderableTopologyNode[]>();
+    const nodeToLevelMap = new Map<RenderableTopologyNode, number>();
+    const nodeCoOrdinatesMap: Map<string, TopologyCoordinates> = new Map<string, TopologyCoordinates>();
+
+    this.findRootNodes(topology, levelToNodesMap, nodeToLevelMap);
+    this.fillNodeAndLevelMaps(levelToNodesMap, nodeToLevelMap);
+    this.fillNodeCoOrdinatesMap(nodeCoOrdinatesMap, levelToNodesMap);
+    this.setNodeCoOrdinates(topology, nodeCoOrdinatesMap);
+  }
+
+  private findRootNodes(topology: RenderableTopology<TopologyNode, TopologyEdge>, levelToNodesMap: Map<number, RenderableTopologyNode[]>, nodeToLevelMap: Map<RenderableTopologyNode, number>) {
+    topology.nodes.forEach(node => {
+      if (node.incoming.length === 0) {
+        nodeToLevelMap.set(node, 0);
+        levelToNodesMap.has(0) ? levelToNodesMap.get(0)?.push(node) : levelToNodesMap.set(0, [node])
+      }
+    });
+  }
+
+  private fillNodeAndLevelMaps(levelToNodesMap: Map<number, RenderableTopologyNode[]>, nodeToLevelMap: Map<RenderableTopologyNode, number>) {
+
+    const rootNodes: RenderableTopologyNode[] = levelToNodesMap.get(0) ?? [];
+    const visited = new Set<string>();
+    let curr = 0;
+
+    while (curr >= 0) {
+      const parent = rootNodes[curr];
+      const level = nodeToLevelMap.get(parent)! + 1;
+
+      parent.outgoing.forEach(edge => {
+        const child = edge.target;
+        if (!visited.has(child.userNode.data?.name!)) {
```

Comment:
If we use node here, we don't even need `!`  here. And this basically says the same thing, if data can be undefined then we shouldn't base it on the data properties

Reply:
Yeah right, updating

Reply:
updated